src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -219,3 +220,343 @@ bool Compiler::optRedundantBranch(BasicBlock* const block)
+            return false;
+        }
+    }
+
+    // In order to optimize we have to be able to determine which predecessors
+    // are correlated exclusively with a true value for block's relop, and which
+    // are correlated exclusively with a false value (aka true preds and false preds).
+    //
+    // To do this we try and follow the flow from domBlock to block; any block pred
+    // reachable from domBlock's true edge is a true pred, and vice versa.
+    //
+    // However, there are some exceptions:
+    //
+    // * It's possible for a pred to be reachable from both paths out of domBlock;
+    // if so, we can't jump thread that pred.
+    //
+    // * It's also possible for us to think a pred is reachable from both paths
+    // when it no longer is; we use fgReachable which was computed upstream and
+    // is not updated as flow is refined. So it may overstate reachability.
+    // To try and mitigate that, we have the main branch loop work in postorder.
+    //
+    // * It's also possible that a pred can't branch directly to a successor as
+    // it might violate EH region constraints. Since this causes the same issues
+    // as an ambiguous pred we'll just classify these as ambiguous too.
+    //
+    // * It's also possible to have preds with implied eh flow to the current
+    // block, eg a catch return, and so we won't see either path reachable.
+    // We'll handle those as ambiguous as well.
+    //
+    // * It's also possible that the pred is a switch; we could handle that
+    // but don't, currently.
+    //
+    // For true preds and false preds we can reroute flow. It may turn out that
+    // one of the preds falls through to block. We would prefer not to introduce
+    // a new block to allow changing that fall through to a jump, so if we have
+    // both a pred that is not a true pred, and a fall through, we defer optimizing
+    // the fall through pred as well.
+    //
+    // This creates an ordering issue, and to resolve it we have to walk the pred
+    // list twice. Classification of preds should be cheap so we just rerun the
+    // reachability checks twice as well.
+    //
+    int               numPreds          = 0;
+    int               numAmbiguousPreds = 0;
+    int               numTruePreds      = 0;
+    int               numFalsePreds     = 0;
+    BasicBlock*       uniqueTruePred    = nullptr;
+    BasicBlock*       uniqueFalsePred   = nullptr;
+    BasicBlock*       fallThroughPred   = nullptr;
+    BasicBlock* const trueSuccessor     = domBlock->bbJumpDest;
+    BasicBlock* const falseSuccessor    = domBlock->bbNext;
+    BasicBlock* const trueTarget        = block->bbJumpDest;
+    BasicBlock* const falseTarget       = block->bbNext;
+
+    for (flowList* pred = block->bbPreds; pred != nullptr; pred = pred->flNext)
+    {
+        BasicBlock* const predBlock = pred->flBlock;
+        numPreds++;
+
+        // We don't do switch updates, yet.
```

Comment:
question: why can't we leave the switches flowing into the original blocks as we do with other AmbiguousPreds?

Reply:
I should just add handling for switches; it's not that hard, just a different kind of update.